### ISSUE TITLE
DiscreteNonParametric should be <: Continuous

### ISSUE THREAD

**Issue by abraunst** [JuliaStats/Distributions.jl#887]
The documentation at https://github.com/JuliaStats/Distributions.jl/blob/master/docs/src/types.md declares that distributions with domain in the reals should be subtypes of `Continuous`. However, `DiscreteNonParameteric` in https://github.com/JuliaStats/Distributions.jl/blob/master/src/univariate/discrete/discretenonparametric.jl is declared as subtype of `Discrete`, violating this rule. This seems a bug (see also my comments at the end of #332).

**Comment by matbesancon** [comment]
Agreed on that.
Sort-of related to https://github.com/JuliaStats/Distributions.jl/issues/771

**Comment by richardreeve** [comment]
This also seems to be related to #916 though that PR actually only adds non-real support. It would be good to fix this `ValueSupport` bug, and I don't see why the `Discrete` subtype has to have `Int` elements at all, it just needs to have a finite (countably infinite?) number of whatever its element type is.

**Comment by mschauer** [comment]
At this point where we need to define `Discrete <: Continuous` , we become aware that we all want to abolish the type params :-D

**Comment by richardreeve** [comment]
Well, I have to admit I'm not thinking of going quite that far! I was imagining `DiscreteSupport <: DiscontinuousSupport <: ValueSupport` and `ContinuousSupport <: ValueSupport`, and allowing DiscreteSupport to allow support over any countable set, whereas ContinuousSupport is uncountable, and Discontinuous is a statement about have Dirac deltas in the density function.

**Comment by DilumAluthge** [comment]
Yeah, I would argue that "discrete" doesn't mean "the support is a subset of integers", but rather "the support is a countable set".

**Comment by mschauer** [comment]
And continuous does not mean "support is a subset of the reals" but should mean perhaps that the distribution has a density (we are not going to model cantor distributions...)

**Comment by richardreeve** [comment]
Does continuous require that support is a `Number` though? That would cover a lot of bases (including complex numbers, `Unitful` numbers, etc.), though we would also need the underlying number type (? I'm not sure there's anything that extracts this at the moment) to be `<: AbstractFloat` I imagine?

**Comment by mschauer** [comment]
Just to state it clearly: The type system needs fixing, but  `DiscreteNonParametric` should not be `<: ContinuousDistribution`

**Comment by richardreeve** [comment]
The solution I proposed last year (to Dirac and all other distributions like this) in #941, #945 and ultimately #951 was to overhaul the whole system and fix all this, but there after several iterations (and months), there were always new minor issues being raised until I had no more time to work on it and so there was never enough enthusiasm to merge it. It’s a shame how much time was wasted.

**Comment by abraunst** [comment]
> The solution I proposed last year (to Dirac and all other distributions like this) in #941, #945 and ultimately #951 was to overhaul the whole system and fix all this, but there after several iterations (and months), there were always new minor issues being raised until I had no more time to work on it and so there was never enough enthusiasm to merge it. It’s a shame how much time was wasted.

For what is worth, I think that yours is the right way. :+1:
